### DISCUSSION TITLE
Pedagogical Interactive Notebooks for Algorithms Implemented in NetworkX for GSoC 2023 Proposal

### DISCUSSION THREAD

**Discussion by whym1here** [networkx/networkx#6448] [General]
I saw this project idea on NetworkX Mentored Projects page. I am really interested in taking this as my GSoC 2023 project. 

**Steps that I would be following for the project if given the opportunity:**
- Study the specific algorithms and its use case in real world (like A* algorithms use in Google Maps or Minimum Spanning Tree use in Water Supply Systems).
- Study its NetworkX api and all its functionalities. In case of any issue I would like to discuss it with my mentor if given the opportunity. 
- Look for some real work application of it with relevant dataset. 
- Create a draft of the notebook and get reviewed with the mentor for the changes or any errors. 
- Help with networkx/networkx/issues/5824 as documentation for this project can be used for Visual map. 

**Few Queries regrading the project:**
- What are the majors algorithms or graph problems that should be the primary focus of the project? 
- I saw 3 different types of notebooks on [NetworkX Notebooks](https://networkx.org/nx-guides/index.html) .These were problem driven notebooks, algorithm driven notebooks and api driven notebooks. What is the expected type of notebook among these? (I am presuming algorithm driven one)
- Some resources on dataset which can used in the above projects. Kaggle does have some datasets and I think we can also generate some networks using Networkx apis but not sure how to approach this problem? 
- What would be the primary medium of communication between Mentor and Contributor?

**Comment by whym1here** [comment]
@MridulS could you answer my queries if possible?

**Reply by MridulS** [reply]
Hi @still-n0thing, sorry for the delay in responding. Was out for the weekend :)

> What are the majors algorithms or graph problems that should be the primary focus of the project?

We usually keep this as part of the exercise for the contributor. If there is something you are specifically interested in please feel free to propose that. You can share your proposal for a quick review before the official proposal deadline so we can suggest changes if required.

> I saw 3 different types of notebooks on [NetworkX Notebooks](https://networkx.org/nx-guides/index.html) .These were problem driven notebooks, algorithm driven notebooks and api driven notebooks. What is the expected type of notebook among these? (I am presuming algorithm driven one)

This is a very flexible project and it depends on what the contributor is interested in exploring. I would suggest to pick up a couple of algorithm classes and review them in detail. Something like https://networkx.org/nx-guides/content/algorithms/euler/euler.html is a nice example. 

> Some resources on dataset which can used in the above projects. Kaggle does have some datasets and I think we can also generate some networks using Networkx apis but not sure how to approach this problem?

For network datasets specifically you can use https://snap.stanford.edu/data/index.html

If you want to look into generating graphs using networkx you can look at https://networkx.org/documentation/latest/reference/generators.html 

> What would be the primary medium of communication between Mentor and Contributor?

We keep majority of communication in the open, like this github discussion thread. We can use email threads when necessary. We also have weekly developer meetings https://scientific-python.org/calendars/ and during the official GSoC period we also have weekly GSoC meetings with contributors who are selected.

Feel free to ping me if you have any other questions :)

**Reply by whym1here** [reply]
@MridulS I will be using this template [PyBaMM Template](https://docs.google.com/document/d/1gER-yFt5_exHEu9Lx-jfrTH8I7QQXci8oDKH-5wgMys/edit?usp=sharing) to draft my proposal. Will share it with you through mail once we have finalized the topics. 

Some of the topics that I am really interested in are:
1. Uses of Minimum Spanning Tree in Metro System Design. (Related to one of my projects) 
2. Use of A star algorithm and Dijkstra Algorithm in finding Single Source Shortest Path in real world. Why do we use A star rather than Dijkstra. Used in different navigation applications. 
3. Use of Directed Graphs for Nondeterministic Finite Automaton(NFA) to Deterministic Finite Automata(DFA) conversion. Used in Scanners or Lexical Analyzers. 
4. Use of Strongly Connected Components in finding groups of people who are more closely related in a huge set of data. Used in recommendation systems for social media applications.

Are these topics relevant or should I think of some other topics?
